packages/nouns-contracts/contracts/governance/NounsDAOLogicV2.sol

```diff
@@ -0,0 +1,835 @@
+            proposal.forVotes = proposal.forVotes + votes;
+        } else if (support == 2) {
+            proposal.abstainVotes = proposal.abstainVotes + votes;
+        }
+
+        receipt.hasVoted = true;
+        receipt.support = support;
+        receipt.votes = votes;
+
+        return votes;
+    }
+
+    /**
+     * @notice Admin function for setting the voting delay
+     * @param newVotingDelay new voting delay, in blocks
+     */
+    function _setVotingDelay(uint256 newVotingDelay) external {
+        require(msg.sender == admin, 'NounsDAO::_setVotingDelay: admin only');
+        require(
+            newVotingDelay >= MIN_VOTING_DELAY && newVotingDelay <= MAX_VOTING_DELAY,
+            'NounsDAO::_setVotingDelay: invalid voting delay'
+        );
+        uint256 oldVotingDelay = votingDelay;
+        votingDelay = newVotingDelay;
+
+        emit VotingDelaySet(oldVotingDelay, votingDelay);
+    }
+
+    /**
+     * @notice Admin function for setting the voting period
+     * @param newVotingPeriod new voting period, in blocks
+     */
+    function _setVotingPeriod(uint256 newVotingPeriod) external {
+        require(msg.sender == admin, 'NounsDAO::_setVotingPeriod: admin only');
+        require(
+            newVotingPeriod >= MIN_VOTING_PERIOD && newVotingPeriod <= MAX_VOTING_PERIOD,
+            'NounsDAO::_setVotingPeriod: invalid voting period'
+        );
+        uint256 oldVotingPeriod = votingPeriod;
+        votingPeriod = newVotingPeriod;
+
+        emit VotingPeriodSet(oldVotingPeriod, votingPeriod);
+    }
+
+    /**
+     * @notice Admin function for setting the proposal threshold basis points
+     * @dev newProposalThresholdBPS must be greater than the hardcoded min
+     * @param newProposalThresholdBPS new proposal threshold
+     */
+    function _setProposalThresholdBPS(uint256 newProposalThresholdBPS) external {
+        require(msg.sender == admin, 'NounsDAO::_setProposalThresholdBPS: admin only');
+        require(
+            newProposalThresholdBPS >= MIN_PROPOSAL_THRESHOLD_BPS &&
+                newProposalThresholdBPS <= MAX_PROPOSAL_THRESHOLD_BPS,
+            'NounsDAO::_setProposalThreshold: invalid proposal threshold'
+        );
+        uint256 oldProposalThresholdBPS = proposalThresholdBPS;
+        proposalThresholdBPS = newProposalThresholdBPS;
+
+        emit ProposalThresholdBPSSet(oldProposalThresholdBPS, proposalThresholdBPS);
+    }
+
+    function _setMinQuorumVotesBPS(uint16 newMinQuorumVotesBPS) external {
+        DynamicQuorumParams memory params = getDynamicQuorumParamsAt(block.number);
+        params.minQuorumVotesBPS = newMinQuorumVotesBPS;
+        _setDynamicQuorumParams(params);
+    }
+
+    function _setMaxQuorumVotesBPS(uint16 newMaxQuorumVotesBPS) external {
+        DynamicQuorumParams memory params = getDynamicQuorumParamsAt(block.number);
+        params.maxQuorumVotesBPS = newMaxQuorumVotesBPS;
+        _setDynamicQuorumParams(params);
```

Comment:
Feels like it may be better to split out validation and events to avoid revalidating and emitting unchanged params. Thoughts?

i.e.

```solidity
function _setMaxQuorumVotesBPS(uint16 newMaxQuorumVotesBPS) external {
    require(msg.sender == admin, 'NounsDAO::_setMaxQuorumVotesBPS: admin only');
    require(
        newMaxQuorumVotesBPS <= MAX_QUORUM_VOTES_BPS_UPPER_BOUND,
        'NounsDAO::_setMaxQuorumVotesBPS: invalid max quorum votes bps'
    );
    DynamicQuorumParams memory params = getDynamicQuorumParamsAt(block.number);

    uint16 oldMaxQuorumVotesBPS = params.maxQuorumVotesBPS;
    params.maxQuorumVotesBPS = newMaxQuorumVotesBPS;

    _writeQuorumParamsCheckpoint(params);

    emit MaxQuorumVotesBPSSet(oldMaxQuorumVotesBPS, newMaxQuorumVotesBPS);
}
```

Reply:
Not thinking of gas savings here. Just seems a bit cleaner to reflect the exact state changes in the events, rather than emitting the same event multiple times in the same block if updating 2+ values.

Reply:
@solimander makes sense, so we will add a task to refactor this code + update the subgraph code to index these individual events as well.

Reply:
🙏

Reply:
Pushed a fix on the smart contract side; subgraph update still TODO

Reply:
subgraph updated